Add support for anchor links

attackant opened the pull request:
Fixes #1039 

This PR does a few things:

- Adds ability to add anchor links to HTML elements in body and text components
- Handles validation and deduplication of IDs
- Fixes issue where non static method was called statically `( new BC_Setup() )->action_init();`
- Fixes array to string conversion error
- Extracts HTML cleanup into multiple methods
- Updates unit tests
- Fixes an edge case where `Apple_News:is_initialized()` never returns true
- Incremental code cleanup and modernization
- Adding some type checking to private methods

attackant committed:
Refactor parser class for stricter type declarations and efficiency

Introduced more strict type declarations in the parser class. This includes making the type of 'format' string explicitly and also ensuring all methods have explicit input and output types. Also, improved some code efficiency by replacing some strpos checks with str_starts_with and str_contains, and handling non-breaking spaces in a more efficient way by using an array in str_ireplace.

attackant committed:
Refactor HTML cleaning in class-parser.php for clarity and improved functionality

The clean_html function of the class-parser.php file has been refactored to improve readability and functionality. New private methods have been introduced to handle specific tasks such as removing empty tags, handling anchor links, root-relative URLs validation, and the conversion of non-breaking spaces to regular spaces. The upgrade has improved the code's efficiency, readability, and maintainability.

attackant committed:
Add namespaces and improve code readability in Class_Body

Several enhancements are introduced to the Class_Body of the Apple Exporter. The changes include addition of namespaces for Theme and DOMElement, explicit definition of return types, and improvement in code readability.

attackant committed:
Refactor class-apple-news.php with strong types and fix bug where is_initialized returns false after saving settings.

The codebase for 'class-apple-news.php' has been significantly refactored to utilize explicit typing for better code readability and maintainability. More specific type hinting is introduced for function parameters and return types, contributing to stricter code standards. A bug, causing the function 'is_initialized' to return false even after saving settings, has also been fixed in this commit.

attackant committed:
Formatting only

Two formatting changes have been made in class-apple-news.php file. Indentation of a comment block was corrected for the $with_padding parameter. Additionally, the statement initializing the $settings variable was reformatted for improved readability.

attackant committed:
Initialize $is_initialized to false and refactor settings check to fix fatal

The $is_initialized static variable in class-apple-news.php is now explicitly initialized to false. The logic for checking if the necessary plugin settings are initialized has also been refactored to improve readability and simplify the boolean expression.

attackant committed:
Improve Apple News push action error handling

Revised error handling in the Apple News 'push' action and corresponding unit test. The code now uses a regular expression for error detection in tests and 'str_contains' instead of 'preg_match' in the actual application.

attackant committed:
Remove anchor link handling from HTML parser

Anchor link handling in the HTML parser has been removed to simplify the code and make it cleaner. Anchor links, which were previously replaced with the global post permalink, will now fall under the regular expression used for managing different link types in the HTML.

attackant committed:
Update anchor tag and corresponding link handling in tests

Anchor tag attributes have been updated from 'name' to 'id' within the test-class-parser test. Corresponding anchor link references have been updated to directly link to the 'id' attribute.

attackant committed:
Formatting, type initialization, readabilty changes

Improved the class-component.php script by correcting namespaces, refining variable initialization, and adjusting type hinting in methods. Added and updated comments to enhance understanding and readability of the code.

attackant committed:
Add TODO, remove type hints

attackant committed:
Refactor link handling in test code

Anchor links are now permitted without auto-conversion. Additionally, the URL detection has been optimized to use the PHP built-in "str_starts_with" function for better performance.

attackant committed:
phpcbf and fix array to string conversion error, add error handling for json_decode

attackant committed:
Fix for non-static method being called statically, misc cleanup

This commit refactors several parts of the codebase for improved readability, performance and adherence to coding standards. Changes include the use of PHP 8's `str_contains()` and `str_starts_with()`, addressing typos in comments, plus removing unnecessary imports. A few adjustments were made to comments and annotations for clearer understanding.

attackant committed:
Roll back some type definitions in public methods

attackant committed:
Cleanup and formatting

attackant committed:
PHPCS fixes

This commit carries out a number of improvements to enhance both the readability and functionality. Firstly, it fixes a comment punctuation issue in 'class-apple-news.php'. Secondly, it substitutes 'json_encode()' with 'wp_json_encode()' in 'class-request.php' to ensure the proper encoding of data in WordPress. Finally, it cleans up the formatting in 'class-export.php', removing unnecessary backslashes and adding parentheses for instantiation of 'BC_Setup' class.

attackant committed:
Add punctuation to comment in class-body.php

This commit adds punctuation to the existing comment in the 'class-body.php' file for clearer readability. It also improves the overall code clarity and standards conformance according to PHP CodeSniffer's rules.

attackant committed:
Add 'id' attribute permission to HTML tags in class-html.php

This commit introduces an update to the allowed_html array in class-html.php, where an 'id' attribute is now allowed for a variety of HTML tags.

attackant committed:
Remove type hinting and refactor methods in class-parser.php

Removed the type hinting from the constructor and parse methods in class-parser.php to allow more flexibility in input types. The HTML formatting is now applied directly within the 'apple_news_parse_html' filter instead of being processed separately, and excess whitespace is removed from the return value of the 'convert_spaces' method.

attackant committed:
Correct TODO comments in class-request.php

Updated the comments of a check on the admin email to reflect that the issue is inconsistency in return value rather than missing return value. This change will provide more accurate information within the comment for future modifications.

attackant committed:
Update allowed HTML tags' attributes in class-component

Adjusted the attributes of the HTML tags allowed in 'class-component.php'. Along with changing 'href' to boolean for 'a' tag, 'id' attributes have now been included for most tags in the 'allowed_html' array. This modification will enable more precise element targeting and manipulation.

attackant committed:
Fix register_spec call

attackant committed:
Introduce HTML id to identifier conversion

Introduced a method to convert HTML 'id' attributes into unique identifiers in the text of a component to support internal anchor links. This includes properly handling potential duplicates and identifiers starting with digits, which are not allowed in Apple News.

attackant committed:
Refactor input types in Apple News class methods

Removed the strict typing for input parameters in `filter_update_post_metadata`, `all_keys_exist`, and `clone_settings` methods. This change includes addition of internal checks for data types within methods to handle possible variations. The new implementation offers more flexibility for input types and includes better error handling for cases when non-array inputs are provided.

attackant committed:
Update duplicate identifier handling logic

Altered how duplicate identifiers are managed in 'class-components.php'. Instead of concatenating a count to duplicate identifiers, the updated code now skips any that already exist. Also, added a TODO comment in 'class-heading.php' to indicate a planned enhancement: adding ID identifiers to anchor links within headings, improving navigation within Apple News structures.

attackant committed:
Remove duplicate identifier handling and improve text flow

Updated the function in 'class-components.php' to remove any duplicate identifiers, improving text flow at all orientations. This change replaces the old method which added an increasing count to duplicate identifiers. In addition, a TODO comment has been added to 'class-heading.php' outlining a future enhancement for adding ID identifiers to anchor links within headings, further enhancing navigation in Apple News structures.

mogmarsh commented on includes/apple-exporter/builders/class-components.php:
```diff
@@ -426,10 +457,69 @@ private function clean_up_components( &$component ) {
+
+				// If this identifier already exists skip it (is a duplicate).
+				if ( isset( $identifiers[ $identifier ] ) ) {
+					return '';
```
do you want to increment the count here? Is the count used for anything? If not, maybe just take that out of the comments.

attackant replied:
Good call, originally I was going to modify the identifiers to be name-1, name-2, but realized that didn't make sense since people wouldn't be able to target them in WordPress. I'll remove.

mogmarsh commented on includes/apple-exporter/class-parser.php:
```diff
@@ -142,72 +138,110 @@ private function parse_markdown( $html ) {
+	 *
+	 * @return string The modified HTML content without empty <a> tags.
+	 */
+	private function remove_empty_tags( string $html ): string {
```
should this be `remove_empty_a_tags` to be a little more descriptive?

mogmarsh commented on includes/apple-exporter/class-parser.php:
```diff
@@ -142,72 +138,110 @@ private function parse_markdown( $html ) {
+	 */
+	private function handle_root_relative_urls( string $html ): string {
+		return preg_replace_callback(
+			'/(<a[^>]+href="\/[^\/"][^"]*")[^>]*>/m',
```
I feel like `[^\/"]` should be `[^\/]`

attackant replied:
I updated the regex slightly to work with both `"/someurl"` and `'/someurl'`

mogmarsh commented on includes/apple-exporter/components/class-body.php:
```diff
@@ -476,12 +482,12 @@ private function dropcap_determination( $theme, $html ) {
-		 * @since 2.4.0
-		 *
 		 * @param bool                  $use_dropcap Whether to apply a dropcap to this paragraph or not.
 		 * @param string                $html The post content to filter.
-		 * @param \Apple_Exporter\Theme $theme The theme whose dropcap options are used.
+		 * @param Theme $theme The theme whose dropcap options are used.
 		 * @param string                $post_id The id of the post whose content we're parsing.
```
do we need to align all the param names now that `\Apple_Exporter\` has been removed?

mogmarsh commented on includes/apple-exporter/components/class-component.php:
```diff
@@ -384,7 +402,7 @@ public function anchor() {
 	 */
 	public function is_anchor_target() {
-		return ! is_null( $this->uid );
+		return '' !== $this->uid;
```
would `! empty()` be good? That would check for `''` and `null`

attackant replied:
Good call, I was using a string before because I was trying to figure out how the variable was changing through the app, but empty is likely a better solution. Will change.

mogmarsh commented on includes/apple-exporter/components/class-component.php:
```diff
@@ -394,7 +412,7 @@ public function is_anchor_target() {
 	 */
 	public function can_be_anchor_target() {
-		return $this->can_be_anchor_target && is_null( $this->uid );
+		return $this->can_be_anchor_target && '' === $this->uid;
```
same

attackant replied:
Done

mogmarsh commented on includes/apple-exporter/components/class-component.php:
```diff
@@ -404,7 +422,7 @@ public function can_be_anchor_target() {
 	 */
 	public function uid() {
-		if ( is_null( $this->uid ) ) {
+		if ( '' === $this->uid ) {
```
same

mogmarsh commented on includes/apple-exporter/components/class-component.php:
```diff
@@ -416,16 +434,17 @@ public function uid() {
 		} else {
-			if ( null === $filename ) {
-				$filename = \Apple_News::get_filename( $source );
+			if ( '' === $filename ) {
```
maybe use `empty()` here too?

mogmarsh commented on includes/class-apple-news.php:
```diff
@@ -1132,7 +1142,7 @@ protected function load_example_themes() {
 	}
 
 	/**
-	 * Verifies that the list of keys provided all exist in the settings array.
+	 * Verifies that the list of keys provided all exists in the settings array.
```
I would go back to `exist` or maybe change it to `Verifies that the keys provided all exist in the settings array.` so there's less confusion.

mogmarsh reviewed (approved):
This looks really good. I made a bunch of comments, but for the most part, they're all suggestions. Feel free to disagree with me and ignore them. 👍

attackant committed:
Simplify identifier handling in class-components.php

Removed counter for duplicate identifiers in 'class-components.php'. The function was adjusted to directly add identifiers into our dictionary without counting, which enhances the text flow. In the future, 'class-heading.php' may include ID identifiers to further improve navigation.

attackant committed:
Modify regular expression for extracting URLs in the function handle_root_relative_urls to handle both single and double quoted urls

attackant committed:
Replace string comparison with empty() function in class-component.php

The commit replaces the string comparison against an empty string checks with the PHP empty() function in class-component.php. This change is expected to make the check more reliable, as it will now also return true for uninitialized variables, null values, and a few other edge cases that weren't handled by the string comparison.

attackant committed:
Fix typo

attackant commented:
Thanks @mogmarsh I went ahead with all of your suggestions. Will merge this after I make another round of update for the Heading component. 🥳

attackant committed:
Start PR

attackant committed:
phpcbf only

attackant committed:
Refactor code for apple exporter header class

The changes encompass various refactoring actions for apple exporter header class. These include implementing proper namespace usage, correctly declaring array data types, specifying return types as well as enhancing PHPDoc comments for methods for better code clarity and readability. Additionally, the future plan to remove duplicate methods across child classes is marked with a TODO comment.

attackant committed:
phpcbf only

attackant committed:
Add ID to heading in tests

In the `test-class-heading.php` and `test-class-parser.php` tests, an ID has been added to the heading in the HTML content. Several test functions have been updated to explicitly specify 'void' as the return type. Apart from these, a small tweak was made to streamline the HTML parsing verification logic.

attackant committed:
Add ID identifier to heading component

Updated the class-heading component to include 'identifier' key for heading IDs. Adjusted regular expression in the build function to capture heading ID value, if set, from HTML input. This addition enhances navigation and provides anchor links to headings.

attackant committed:
phpcbf formatting

attackant committed:
Fix regex to so IDs are optional and update matches variables. Thank you phpunit, you are my only hope.

attackant committed:
Merge pull request #1045 from alleyinteractive/issue-1039-b

Issue 1039 Part 2
